feat: Implement Mustache template support for search queries

Bharathi-Kanna opened the pull request:
### Description
This PR implements backend support for Mustache templates in search queries using OpenSearch's native 
ScriptService

### Key Changes:

#### Integrated Mustache Templating

- Modified SearchRelevancePlugin.javato capture and pass ScriptService
- Updated SearchRequestBuilder.java to compile and execute Mustache templates
- Added automatic detection: queries with {{ use Mustache, others use legacy %SearchText%

#### Maintained Backward Compatibility

- All existing %SearchText% queries continue to work without changes
- No data model modifications
- No migration required

#### Example Usage
Mustache Template (New):
```
{
  "query": {
    "match": { "title": "{{queryText}}" }
  }
}
```
Legacy Placeholder (Still Works):
```
{
  "query": {
    "match": { "title": "%SearchText%" }
  }
}
```
Both syntaxes work side-by-side. The system auto-detects which one to use.

**Current Limitation**: Single-parameter queries only (matches existing QuerySetEntry as List<String>)
_**Multi-parameter support (e.g., {{brand}}, {{category}}) requires data model changes and will be addressed in a future PR.**_

### Issues Resolved
Relates to [#41](https://github.com/opensearch-project/search-relevance/issues/41)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Bharathi-Kanna committed:
feat: Implement Mustache template support for search queries

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

epugh reviewed (commented):
Looks like you are on the right path..

epugh commented:
Looks like good progress.   We still need some tests for handling it..   Let me know how I can help!

Bharathi-Kanna committed:
feat: Add comprehensive test coverage for Mustache template support

- Added 4 test methods to SearchRequestBuilderTests.java
  - testMustacheWithNullScriptService: Validates error handling
  - testLegacyWildcardStillWorks: Confirms backward compatibility
  - testDetectionLogicUsesLegacyForNonMustache: Verifies detection logic
  - testHybridSearchMustacheDetection: Tests hybrid query support
- Updated CHANGELOG.md with feature entry
- All tests passing (13/13 SearchRequestBuilderTests)

Addresses maintainer feedback on PR #342

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna committed:
Merge branch 'main' into feature/mustache-template-support

Signed-off-by: Bharathi Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna committed:
Merge branch 'upstream/main' into feature/mustache-template-support

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna committed:
Add support for multiple Mustache variables via QuerySetEntry customFields

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna committed:
Fix Javadoc mismatches and clean up CHANGELOG entries

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna committed:
Merge remote-tracking branch 'upstream/main' into feature/mustache-template-support

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

# Conflicts:
#	CHANGELOG.md

Bharathi-Kanna committed:
Revert LLM judgment cache-key to upstream JSON format

buildCacheKey had been changed to a newline-delimited format
(queryText#\nkey:value), which silently invalidated existing judgment
cache entries on upgrade — the key is used for both writes and reads, so
entries stored in the released JSON format (queryText#{json}) would no
longer be found, forcing needless LLM recomputation. The change also
contradicted the method's own Javadoc and QuerySetEntry.parseLegacyQueryText,
which both expect the JSON format, and was unrelated to the Mustache
feature (SearchRequestBuilder never touches buildCacheKey).

Restore the upstream JSON serialization via OBJECT_MAPPER (already a
dependency in this file). No test changes.

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna committed:
Test Mustache rendering scenarios and fix the multi-variable IT fixture

Add MustacheTemplateRenderTests, which exercises the real SearchRequestBuilder
render path against a live ScriptService backed by the core MustacheScriptEngine
(the existing SearchRequestBuilderTests only cover the legacy %SearchText% and
null-ScriptService error paths). It deterministically asserts single-variable and
multi-variable substitution, JSON escaping of special characters, empty rendering
of missing variables, and rejection of Mustache partials ({{>...}}) — the latter
verifying OpenSearch core's partial-resolution guard is present.

Pull in lang-mustache-client as a test dependency so the engine is available to
unit tests (the node provides it at runtime).

Fix the multi-variable search-configuration fixture: it referenced an unprovided
{{category_filter}} variable and applied a term filter on the analyzed 'category'
field. Reference the supplied {{category}} custom field and target the
'category_filter' keyword field, using an optional should clause so results are
driven by the title match (the ESCI sample data is not category-filtered).

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna committed:
Fail experiment gracefully when a search template cannot be built

A search configuration whose query is an invalid Mustache template — most
notably one containing an unsupported partial ({{>...}}), which OpenSearch
core now rejects at compile time — caused SearchRequestBuilder to throw
synchronously inside ExperimentTaskManager.executeVariantAsync, before
client.search was invoked. That path never called completeVariantFailure(),
so ExperimentTaskContext.remainingVariants never reached zero and the
experiment hung in PROCESSING indefinitely.

Wrap the request build in a try/catch that routes the failure through the
same handleSearchFailure path used for search errors, so the variant is
counted, the failure is recorded, and the experiment reaches COMPLETED.

Add a regression IT (partial-template search config -> experiment COMPLETED)
plus the supporting fixture and helper.

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>

Bharathi-Kanna committed:
Rename Mustache query variable to {{queryText}} for consistency

The query text was exposed to Mustache templates as {{query_string}}, which did
not match the query set field name (queryText). Custom fields are already
referenced by their field names (for example, {{category}}), so the query text
now follows the same convention and is available as {{queryText}}. Also updates
the processMustacheTemplate Javadoc to document customFields.

Addresses review feedback on #342.

Signed-off-by: Bharathi-Kanna <99189546+Bharathi-Kanna@users.noreply.github.com>
